Access control: Add permissions to admins/users based on the new access control system

vtorosyan opened the pull request:
**What this PR does / why we need it**:

Introduces the first batch of permissions using the new access control system with fallback to the current `reqGrafanaAdmin` mechanism.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana-enterprise/issues/1091

**Special notes for your reviewer**:

vtorosyan committed:
Add fine-grained permissions to admins/users endpoints

vtorosyan commented:
@gamab @alexanderzobnin this is draft, but feel free to have a look to see if anything is off already.

vtorosyan commented on pkg/services/accesscontrol/seeder/roles.go:
```diff
@@ -0,0 +1,68 @@
+		},
+	},
+	{
+		Name:    "roles:adminUsers:viewer",
```
After adding these predefined roles, I am really unsure about the difference between built in roles and predefines roles. Aren't these kind-of the same logical concept?

vtorosyan commented on pkg/services/accesscontrol/seeder/roles.go:
```diff
@@ -0,0 +1,68 @@
+		Permissions: []accesscontrol.Permission{
+			{
+				Permission: "users.password.update",
+				Scope:      "*",
```
It does not seem that evaluator supports `*` now. And currently we are forwarding `users:[id]` with the request and the evaluator is going to check against that specific scope. 

That said, probably I need to add support of `*` in the evaluator.

vtorosyan replied:
Ah I might be wrong actually, it does seem evaluator does this
```
rule, err := glob.Compile(dbScope, ':', '/')
			if err != nil {
				return false, err
			}

			match = rule.Match(s)
			if match {
				break
			}
```

vtorosyan commented:
Since we are evaluating architecture changes, I will block this PR until https://github.com/grafana/grafana-enterprise/issues/1208 this one is done/clear.

vtorosyan commented on pkg/services/accesscontrol/seeder/roles.go:
```diff
@@ -0,0 +1,88 @@
+
+import "github.com/grafana/grafana/pkg/services/accesscontrol"
+
+var builtInRoles = []accesscontrol.RoleDTO{
```
I think it's worth to have a discussion about renaming built-in roles to predefined roles (I will be fine with also renaming predefined roles to built-in roles). But after adding these two new roles, I can see that logically built-in roles and predefined roles are essentially the same concept. cc @alexanderzobnin @gamab

vtorosyan commented on pkg/services/accesscontrol/database/database.go:
```diff
@@ -16,8 +16,10 @@ import (
+// SQLStore is not injected directly to make dependency between AccessControlStore and Manager explicit.
 type AccessControlStore struct {
-	SQLStore *sqlstore.SQLStore `inject:""`
+	SQLStore *sqlstore.SQLStore
```
Dependencies are a bit hard to manage here. While initially I was hoping that I could skip registering access control store as a service, but I run into too many problems with tests, when running migrations. I eventually decided to not inject the SQLStore and init it from the manager, but keep the accesscontrolstore as a service.

vtorosyan commented on pkg/services/accesscontrol/database/database.go:
```diff
@@ -138,6 +140,7 @@ func (ac *AccessControlStore) CreateRoleWithPermissions(ctx context.Context, cmd
 		createRoleCmd := accesscontrol.CreateRoleCommand{
 			OrgID:       cmd.OrgID,
 			UID:         cmd.UID,
+			Version:     cmd.Version,
```
`version` was missing when creating roles.

vtorosyan commented on pkg/services/accesscontrol/models.go:
```diff
@@ -186,3 +186,22 @@ func (p RoleDTO) Role() Role {
 }
+
+const (
+	// Permission actions
```
There is a trade-off between declaring these as constants vs not. I think having them declared in one places is a huge plus for keeping it as a source of truth and documentation.

sakjur replied:
Makes sense. I'm not a fond of globals and having to extend the mental lookup table to include a bunch of constants, but yeah, this might prevent more bugs than it'll introduce.

vtorosyan committed:
Fix linter

vtorosyan committed:
Merge branch 'master' into vt/accesscontrol-endpoints

sakjur commented on pkg/services/accesscontrol/database/database_mig.go:
```diff
@@ -14,14 +14,14 @@ func AddAccessControlMigrations(mg *migrator.Migrator) {
 
 	//-------  indexes ------------------
-	mg.AddMigration("add unique index permission.role_id", migrator.NewAddIndexMigration(permissionV1, permissionV1.Indices[0]))
+	mg.AddMigration("add unique index role_id,permission,scope", migrator.NewAddIndexMigration(permissionV1, permissionV1.Indices[0]))
```
I think you can skip dropping the old index, this has never been released and is behind a feature flag.

sakjur commented on pkg/services/accesscontrol/seeder/roles.go:
```diff
@@ -0,0 +1,88 @@
+		},
+	},
+	{
+		Name:    "roles:adminUsers:viewer",
```
We should have a prefix for all of our predefined roles that we can instruct/prevent users from creating via the APIs.
I'd also suggest making all "manage users" roles sort under `users`, regardless of if the actions are intended for administration or end-users. We're probably going to end up sorting these alphabetically when presenting to users, so I think it's rather important that the alphabetical ordering groups related concepts near each other.
Finally, I suggest that we call it "view" rather than "viewer", or something else like "read" or "list". Naming is hard, but keeping words to their base forms seems reasonable to me. 

The prefix thing seems like a mistake not to do, everything else I'll yield on any amount of resistance, it's my reasoning behind how I decided to name the above `grafana:builtin:users:read:self` (which, for the record, isn't a very good name either (the prefix is too long, so it risks being the only thing users see in a list), so I wouldn't use it as a template).

With those modifications we'd end up with something like the following:

```
gf:users:self:view
gf:users:admin:view
gf:users:admin:edit
```

vtorosyan replied:
I agree with reasoning 👍 I'd also vote for `read` instead of `view` , since it's more known as a term in the IAM world + view is harder to reason about. 

Also the prefix makes sense, have not thought that we would need to apply some prevention logic for these roles. Since it's a convention we want to stick, I would say `gf` prefix works just fine and will save us few bytes of space in the storage.

alexanderzobnin replied:
`read` makes more sense to me. `list` fits well for collections, but I'm not sure we want to differentiate `list` (when user allowed to list resources, but not read it, like you can list users, but cannot read user profile) from `read`.

alexanderzobnin replied:
In our case users search returns items with lots of info about users, so `read` makes more sense than `list` I think.

sakjur replied:
Agreed, list makes sense when there's a value to differentiating accessing something and knowing _that_ it exists. Generally speaking, I think we should avoid that to keep things simple 🙂

sakjur commented on pkg/services/accesscontrol/seeder/roles.go:
```diff
@@ -0,0 +1,88 @@
+			},
+			{
+				Permission: accesscontrol.ActionUsersCreate,
+				Scope:      accesscontrol.ScopeUsersAll,
```
Creating users should probably not be scoped.

sakjur commented on pkg/services/accesscontrol/seeder/seeder.go:
```diff
@@ -14,32 +14,17 @@ type seeder struct {
 	"grafana:builtin:users:read:self": {
 		"Viewer",
 	},
+	"roles:adminUsers:viewer": {
+		"Admin",
+	},
+	"roles:adminUsers:editor": {
+		"Admin",
+	},
```
These are Grafana Admin level permissions, not Admin. I think there's a constant for "Grafana Admin" somewhere.

```suggestion
	"roles:adminUsers:viewer": {
		"Grafana Admin",
	},
	"roles:adminUsers:editor": {
		"Grafana Admin",
	},
```

sakjur commented on pkg/api/api.go:
```diff
@@ -442,6 +434,23 @@ func (hs *HTTPServer) registerRoutes() {
 		adminRoute.Get("/ldap/status", routing.Wrap(hs.GetLDAPStatus))
 	}, reqGrafanaAdmin)
 
+	// Administering users
```
There are a bunch of these endpoints which I missed when going over this, we'll have to make sure to extract the other `admin/users` endpoints from further up in the code as well.

vtorosyan replied:
Yeah I went through few of them, will double check if I missed anything!

sakjur reviewed (commented):
Left some quick comments, I think the changes you've done here makes sense 👍

vtorosyan committed:
Update pkg/services/accesscontrol/seeder/seeder.go

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

vtorosyan committed:
Merge branch 'master' into vt/accesscontrol-endpoints

vtorosyan committed:
Merge branch 'vt/accesscontrol-endpoints' of github.com:grafana/grafana into vt/accesscontrol-endpoints

vtorosyan committed:
Consistent naming for predefined roles

vtorosyan committed:
Allow to skip scope from permission and reflect that in the evaluator

vtorosyan commented on pkg/services/accesscontrol/evaluator/evaluator.go:
```diff
@@ -30,10 +29,14 @@ func Evaluate(ctx context.Context, ac accesscontrol.AccessControl, user *models.
 		var match bool
 		for dbScope := range dbScopes {
+			// If there is no scope requires, allow by default
+			if len(dbScope) == 0 {
```
This is an explicit check to break the loop when the scope is not defined at all, which by default assumes "don't require scope check"

vtorosyan commented:
@sakjur I've updated the PR with latest changes from master and also addressed few of the comments. Would be great if you can take it from here 🙇

vtorosyan commented on pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go:
```diff
@@ -5,29 +5,97 @@ import (
 var builtInRolesMap = map[string]accesscontrol.RoleDTO{
-	"grafana:builtin:users:read:self": {
-		Name:    "grafana:builtin:users:read:self",
+	"gf:users:self:read": {
```
Now that I look at this, maybe `grafana:roles:users:self:read` would be better, 11 bytes more, but more descriptive/clear 🤔 I can imagine someone would want to prefix their roles with `gf`

sakjur replied:
> 11 bytes more

text compresses well 🤷

I am more concerned about listings where you have a list of all roles and the display of the name is capped at some point, but yeah, +/- 11 characters shouldn't make or break it, and having it spelled out is nice.

sakjur committed:
Minor style updates

sakjur committed:
Add access control checks for /api/users

It's the other half of /api/admin/users, they're both intended for Grafana Admins

sakjur committed:
Remove check for empty scope

The first for-loop already handles the case of "no scopes are necessary"

sakjur committed:
Add new scopes to predefined roles

sakjur committed:
Move roles to domain package

vtorosyan commented on pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:
```diff
@@ -41,10 +41,10 @@ func (ac *OSSAccessControlService) Evaluate(ctx context.Context, user *models.Si
 	permissions := make([]*accesscontrol.Permission, 0)
 	for _, legacyRole := range roles {
-		if builtInRoleNames, ok := builtInRoleGrants[legacyRole]; ok {
+		if builtInRoleNames, ok := accesscontrol.PredefinedRoleGrants[legacyRole]; ok {
```
```suggestion
		if predefinedRoleNames, ok := accesscontrol.PredefinedRoleGrants[legacyRole]; ok {
```

vtorosyan commented on pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:
```diff
@@ -41,10 +41,10 @@ func (ac *OSSAccessControlService) Evaluate(ctx context.Context, user *models.Si
 	for _, legacyRole := range roles {
-		if builtInRoleNames, ok := builtInRoleGrants[legacyRole]; ok {
+		if builtInRoleNames, ok := accesscontrol.PredefinedRoleGrants[legacyRole]; ok {
 			for _, builtInRoleName := range builtInRoleNames {
```
```suggestion
			for _, predefinedRoleName := range predefinedRoleNames {
```

vtorosyan commented on pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:
```diff
@@ -41,10 +41,10 @@ func (ac *OSSAccessControlService) Evaluate(ctx context.Context, user *models.Si
 			for _, builtInRoleName := range builtInRoleNames {
-				builtInRole := getBuiltInRole(builtInRoleName)
-				if builtInRole == nil {
+				builtInRole, exists := accesscontrol.PredefinedRoles[builtInRoleName]
```
```suggestion
				predefinedRole, exists := accesscontrol.PredefinedRoles[predefinedRoleName]
```

vtorosyan commented on pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:
```diff
@@ -41,10 +41,10 @@ func (ac *OSSAccessControlService) Evaluate(ctx context.Context, user *models.Si
+				if !exists {
 					continue
 				}
 				for _, p := range builtInRole.Permissions {
```
```suggestion
				for _, p := range predefinedRole.Permissions {
```

vtorosyan commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,115 @@
+
+import "github.com/grafana/grafana/pkg/models"
+
+var PredefinedRoles = map[string]RoleDTO{
```
Are we going to rename the database table to `predefined_roles` and reflect that in the code if relevant? In other words, should talk about "predefined roles" and not "built-in roles" from now on?

sakjur replied:
No, unfortunately those are two slightly different concept, the built-in roles are the Viewer/Editor/Admin trio (plus Grafana Admin) and predefined roles are just a list of roles that _we_ define. It'd be nice with a better/clearer naming here, but unfortunately, Viewer/Editor/Admin are already called "role" in a lot of places.

We could rename the database table to something with "org role" maybe, because it's really what role the user has access to organization wide?

sakjur commented on pkg/api/api.go:
```diff
@@ -154,16 +159,17 @@ func (hs *HTTPServer) registerRoutes() {
+			const userIDScope = `users:{{ index . ":id" }}`
+			usersRoute.Get("/", authorize(reqGrafanaAdmin, accesscontrol.ActionUsersRead, accesscontrol.ScopeUsersAll), routing.Wrap(SearchUsers))
+			usersRoute.Get("/search", authorize(reqGrafanaAdmin, accesscontrol.ActionUsersRead, accesscontrol.ScopeUsersAll), routing.Wrap(SearchUsersWithPaging))
+			usersRoute.Get("/:id", authorize(reqGrafanaAdmin, accesscontrol.ActionUsersRead, userIDScope), routing.Wrap(GetUserByID))
```
Is this a change in access level?

No, for two reasons:
- These endpoints do not call something that isn't in the background called by the `/user` endpoint accessible to users.
- We don't actually grant users access to `users:read` for `users:<id>` but only for the literal `users:self`. This might change, I think it'd be reasonable to change that, but that means that _at the moment_ this is not a change in scope.

vtorosyan commented on pkg/services/accesscontrol/models.go:
```diff
@@ -75,3 +75,34 @@ func (p RoleDTO) Role() Role {
+	ActionUsersAuthTokenUpdate = "users.authtoken:update"
+	// We can ignore gosec G101 since this does not contain any credentials
+	// nolint:gosec
+	ActionUsersPasswordUpdate    = "users.password.update"
```
```suggestion
	ActionUsersPasswordUpdate    = "users.password:update"
```

vtorosyan commented on pkg/services/accesscontrol/models.go:
```diff
@@ -75,3 +75,34 @@ func (p RoleDTO) Role() Role {
+	ActionUsersCreate            = "users:create"
+	ActionUsersEnable            = "users:enable"
+	ActionUsersDisable           = "users:disable"
+	ActionUsersPermissionsUpdate = "users.permissions.update"
```
```suggestion
	ActionUsersPermissionsUpdate = "users.permissions:update"
```

vtorosyan commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,115 @@
+	usersAdminEdit: {
+		Name:    usersAdminEdit,
+		Version: 1,
+		Permissions: []Permission{
```
It seems this one is missing users read and teams read? 
```
{
				Permission: ActionUsersRead,
				Scope:      ScopeUsersAll,
			},
			{
				Permission: ActionUsersTeamRead,
				Scope:      ScopeUsersAll,
			},
```

sakjur replied:
🤔 Oh yeah, we talked about making it so that edit should inherit read 🤦. Ugh. I don't want to duplicate this everywhere, it's too fragile to manually inherit permissions that way. Will have to rebuild some things to accomodate for that.

vtorosyan commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,115 @@
+		Version: 1,
+		Permissions: []Permission{
+			{
+				Permission: "users:read",
```
```suggestion
				Permission: ActionUsersRead,
```

sakjur replied:
Oh, I forgot to remove this thing. It's not used anywhere, so I'm removing it instead of applying these changes.

vtorosyan commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,115 @@
+				Scope:      ScopeUsersSelf,
+			},
+			{
+				Permission: "users.tokens:list",
```
```suggestion
				Permission: ActionUsersAuthTokenList,
```

vtorosyan commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,115 @@
+				Scope:      ScopeUsersSelf,
+			},
+			{
+				Permission: "users.teams:read",
```
```suggestion
				Permission: ActionUsersTeamRead,
```

sakjur committed:
Address review comments

sakjur committed:
Update pkg/services/accesscontrol/models.go

Co-authored-by: Vardan Torosyan <vardants@gmail.com>

sakjur committed:
Remove unused role

sakjur committed:
Update pkg/services/accesscontrol/models.go

Co-authored-by: Vardan Torosyan <vardants@gmail.com>

sakjur committed:
Merge branch 'master' into vt/accesscontrol-endpoints

vtorosyan committed:
Fix middleware to update templated scopes in the consequent requests

vtorosyan commented:
I've recorded a demo which goes over the impact of the changes in this PR, feel free to [check the recording in the design doc](https://docs.google.com/document/d/1BaPsYtBfdYipaBJtiClE6aFU2o1ueKMqkqjWtnrZG1c/edit#heading=h.4kzfookpmy61)

Note that the demo covers the whole fine-grained access control, but it is based on the endpoints which have been changed within this PR.

vtorosyan committed:
Fix failing tests

vtorosyan commented on pkg/services/accesscontrol/models.go:
```diff
@@ -62,3 +62,34 @@ func (p RoleDTO) Role() Role {
+	// Permission actions
+
+	ActionUsersRead     = "users:read"
+	ActionUsersWrite    = "users:write"
```
Since we have users:create for creating new users, do you think users:write is clear for editing? 🤔  Should we rename it to users:update to be consistent with `users.password:update` ?

sakjur replied:
🤔 I think `users:write` is clear for editing. I think the `users.password:update` could be changed to `users.password:write`, I think the reason why I went with update there is because passwords cannot be read, so a `users.password:read` would never make sense.

vtorosyan replied:
Sounds good to me.

sakjur committed:
Extend admin edit scope with new permissions

sakjur committed:
Add consistency check tests for roles.go

sakjur committed:
Merge branch 'master' into vt/accesscontrol-endpoints

sakjur committed:
Merge branch 'master' into vt/accesscontrol-endpoints

# Conflicts:
#	pkg/services/accesscontrol/evaluator/evaluator.go
#	pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go
#	pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go

sakjur committed:
Permission->Action

sakjur committed:
Fix roleGrafanaAdmin

alexanderzobnin commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,113 @@
+var PredefinedRoles = map[string]RoleDTO{
+	// TODO: Add support for inheritance between the predefined roles to
+	// make the admin ⊃ editor ⊃ viewer property hold.
+	usersAdminRead: {
```
This is more related to the frontend and maybe outside this scope, but just want to mention. In the old authorization model we render side menu based on built-in roles. Now we have to support fine grain access control and in order to properly render the menu, we should define roles for each menu item, like `Server Admin -> Users`, `Server Admin -> Orgs`, `Server Admin -> Settings`, etc. That's related to code here

https://github.com/grafana/grafana/blob/b0470c84a5eac8b7871e5a2e4711cadc5e67f93d/pkg/api/index.go#L301-L307

sakjur replied:
Yes, I'm working on that in grafana/grafana-enterprise#1251, but it's out of scope for this PR which "just" covers the API (to do that, I _think_ we're also going to have to change the middlewares for the frontend endpoints) 🙂

alexanderzobnin replied:
So in order to make grafana UI behave properly, we have to add that permission. Here or in the other PR.

alexanderzobnin commented on pkg/services/accesscontrol/roles.go:
```diff
@@ -0,0 +1,113 @@
+// PredefinedRoleGrants specifies which organization roles are assigned
+// to which set of PredefinedRoles by default. Alphabetically sorted.
+var PredefinedRoleGrants = map[string][]string{
+	RoleGrafanaAdmin: {
```
That's great, we can split GrafanaAdmin role into a set of small fine-grained roles.

alexanderzobnin reviewed (approved):
Looks good!
